Approach.md

```diff
@@ -0,0 +1,102 @@
+Read-only interactions such as query and stream can happen concurrently without issue.
+
+## rrocker-cli:
+The CLI will need to be run once per command, so scheduling multiple tasks requires multiple invocations.
+
+There'll be zero command line switches and as such it'll be hardcoded to connect to `rrockerd` running on localhost.
+
+## rrocker-lib:
+No attention will be paid to backwards compatibility of the API, meaning no versioning or abstractions.
+
+Human friendliness of the API is a distant afterthought, as such full UUIDs will be required to interact with tasks.
+
+Resource limits will be exposed 1-to-1 with the underlying cgroups API leaving it up to the `rrocker-cli` to expose it in a human friendly way.
+# Security
+Security is naturally an important aspect of the implementation however as my experience with crypto is limited I'll attempt to follow best practices and use the default configuration of OpenSSL and rustTLS wherever applicable.
+In an actual production system you'd have a security expert decide the configuration based on a threat model.
+
+## Authentication
+Authentication will be done with mTLS according to the challenge rules.
+
+The following certificates will be created:
+- Root Certificate Authority (CA). This is our root of trust and should be stored on an air-gapped system that's only used to sign/revoke the server/client CAs.
+- Separate server and client CAs. These are used to sign/revoke server and client certs. Furthermore they're separated such that an infrastructure team is able to own their CA and deploy new servers independently. These CAs also introduce an indirection such that servers and clients don't need to know about all clients and servers but can simply verify an identity has been signed by the server/client CA.
+- Server_1 cert used by the server to auth with the clients
+- Client_1 and Client_2 certs used to demo user's only being able to see their own tasks
+- Admin_1 cert used to demo admin's being able to see all tasks
+- An untrusted CA + client cert to demo unauthorized client's can't connect
+
+Private keys will be generated with OpenSSL using the prime256v1 ECDH curve as recommended by mozilla: https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility.
+
+## Authorization
+The authorization scheme will be super simple solution where an authenticated user either is an admin or regular user.
+A regular user will only be able to perform actions on it's own tasks while an admin can interact with **all** tasks.
+
+This distinction will be done by looking at the Common Name of the client certificate, `CN=client` for clients and `CN=admin` for admins.
+The public key of a cert will be used as the client id despite the drawback that a single client would have to share the same cert across multiple servers against best practices. 
```

Comment:
> The public key of a cert will be used as the client id despite the drawback that a single client would have to share the same cert across multiple servers against best practices. 

Sorry, I'm not sure I follow what the drawback / best practice being violated is here?

Reply:
My understanding of best practices are to keep private keys local and never transfer them between systems as there are risk this might be done in an insecure way.
In a similar vein it's best practices regarding ssh keys for development where it's recommended to generate a unique ssh key for each system you expect to push code from.
This helps with damage control should one of the systems be compromised as you'll only need to revoke one key and investigate commits made with that key.

As such the drawback here is that this corner cutting forces the client to share the same private key across multiple servers in order to access the same group of tasks from multiple servers since different key pairs will see a set of tasks unique to that set of keys.

See my answer to your below question for what I'd recommend for an actual production system.